src/tinygp/kernels/quasisep.py

```diff
@@ -674,78 +683,230 @@ def init(
-        # also have an expression for this (their V_{ij}), but I prefer to use
-        # this since it is probably roughly just as fast to compute, and it is
-        # strictly real-valued.
-        f = 2 * ((np.arange(2 * p) // 2) % 2) - 1
-        x = f * jnp.append(alpha_ext, jnp.zeros(p - 1))
-        params = jnp.stack([np.roll(x, k)[::2] for k in range(p)], axis=0)
-        params = jnp.linalg.solve(
-            params, 0.5 * sigma**2 * jnp.eye(p, 1, k=-p + 1)
-        )[:, 0]
-        stn_ = []
-        for j in range(p):
-            stn_.append([jnp.zeros(()) for _ in range(p)])
-            for n, k in enumerate(range(j - 2, -1, -2)):
-                stn_[-1][k] = (2 * (n % 2) - 1) * params[j - n - 1]
-            for n, k in enumerate(range(j, p, 2)):
-                stn_[-1][k] = (1 - 2 * (n % 2)) * params[n + j]
-        stn = jnp.array(list(map(jnp.stack, stn_)))
+        # find acf
+        arroots = CARMA.roots(jnp.append(alpha, 1.0))
+        acf = CARMA.carma_acf(arroots, alpha, beta * sigma)
+        # masks for selecting entries in matrixes
+        real_mask = jnp.where(arroots.imag == 0.0, jnp.ones(p), jnp.zeros(p))
+        complex_mask = -real_mask + 1
+        complex_idx = jnp.cumsum(-real_mask + 1) * complex_mask
+        complex_select = complex_mask * complex_idx % 2
+
+        # compute obsmodel
+        om_real = jnp.sqrt(jnp.abs(acf.real))
+        a, b, c, d = (
+            2 * acf.real * complex_mask,
+            2 * acf.imag * complex_mask,
+            -arroots.real * complex_mask,
+            -arroots.imag * complex_mask,
+        )
+        c2 = jnp.square(c)
+        d2 = jnp.square(d)
+        s2 = c2 + d2
+        h2_2 = d2 * (a * c - b * d) / (2 * c * s2 + eta * real_mask)
+        h2 = jnp.sqrt(h2_2)
+        h1 = (c * h2 - jnp.sqrt(a * d2 - s2 * h2_2)) / (d + eta * real_mask)
+        om_complex = jnp.array([h1, h2])
+        obsmodel = (om_real * real_mask) + jnp.ravel(om_complex)[
+            ::2
+        ] * complex_mask
 
         return cls(
-            sigma=sigma,
             alpha=alpha,
             beta=beta,
-            roots=roots,
-            proj=proj,
-            proj_inv=proj_inv,
-            stn=stn,
+            sigma=sigma,
+            arroots=arroots,
+            acf=acf,
+            real_mask=real_mask,
+            complex_mask=complex_mask,
+            complex_select=complex_select,
+            obsmodel=obsmodel,
+        )
+
+    @classmethod
+    def from_fpoly(
+        cls, alpha_fpoly: JAXArray, beta_fpoly: JAXArray, beta_mult: JAXArray
+    ) -> "CARMA":
+        """Construct a CARMA kernel using the roots of the characteristic polynomials
+
+        The roots can be re-parameterized as the coefficients of a product
+        of quadratic equations each with the second-order term set to 1. The
+        input for this constructor are said coefficients. See Equation 30 in
+        the paper linked above for a reference.
```

Comment:
It's probably better to directly link to the paper here, instead of referencing to above.

Reply:
OK, fixed it.